Insights Scripts

wbeard opened the pull request:
Just opening a draft as a check-in point before we meet!

wbeard committed:
Add insights versions of scripts:

- postDeploysAndCommitsToInsights()
- updateDeployToInsights()

These call underlying shell scripts to execute

sbhaskara-ca commented on vars/insights/postCommits.sh:
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
+_CURRENT_BUILD_COMMIT=HEAD
+_DEPLOY_ID=423575567561
```
Is this OID for the VSMDeploy? How do you expect to get this?

sbhaskara-ca replied:
Why not do the same logic of finding the DeployID based off the current BuildID

sbhaskara-ca commented on vars/insights/postDeploys.sh:
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+
+_BUILD_COMPONENT=423575445565
```
How would we get the Component OID?

wbeard replied:
We could switch to use a Component name instead, but would mean that if the component were ever renamed, there would have to be a corresponding change here.

I think when we discussed today, we settled on doing this in another PR which I'm up for if it helps simplify onboarding!

sbhaskara-ca commented on vars/postCommitsToInsights.groovy:
```diff
@@ -0,0 +1,79 @@
+def call(Map config = [:]) {
+  sh """
```
@wbeard we do not need to this; we can keep the Shell script as is in the `resources` directory and we can execute it from here. that way its easy to maintain and also testable.

sbhaskara-ca commented on vars/postDeployAndCommitsToInsights.groovy:
```diff
@@ -0,0 +1,18 @@
+        "CURRENT_BUILD_COMMIT=${config.CurrentBuildCommit}",
+        "DEPLOY_BUILD_ID=${config.BuildId}",
+        "DEPLOY_COMPONENT=${config.BuildComponentObjectId}",
+        "DEPLOY_END_TIME=${config.BuildFinishTime}",
```
Doc to be added about the params and formats

sbhaskara-ca commented on vars/postDeployAndCommitsToInsights.groovy:
```diff
@@ -0,0 +1,18 @@
+        "DEPLOY_IS_SUCCESSFUL=${config.BuildIsSuccessful}",
+        "DEPLOY_MAIN_REVISION=${config.CurrentBuildCommit}",
+        "DEPLOY_START_TIME=${config.BuildStartTime}",
+        "GIT_REPO_LOC=${config.GitRepoLoc}",
```
Ensure if we can handle both relative and absolute paths

sbhaskara-ca commented on vars/postDeployAndCommitsToInsights.groovy:
```diff
@@ -0,0 +1,18 @@
+        "DEPLOY_BUILD_ID=${config.BuildId}",
+        "DEPLOY_COMPONENT=${config.BuildComponentObjectId}",
+        "DEPLOY_END_TIME=${config.BuildFinishTime}",
+        "DEPLOY_IS_SUCCESSFUL=${config.BuildIsSuccessful}",
```
Call it out as optional

sbhaskara-ca commented on vars/postDeployAndCommitsToInsights.groovy:
```diff
@@ -0,0 +1,18 @@
+        "DEPLOY_COMPONENT=${config.BuildComponentObjectId}",
+        "DEPLOY_END_TIME=${config.BuildFinishTime}",
+        "DEPLOY_IS_SUCCESSFUL=${config.BuildIsSuccessful}",
+        "DEPLOY_MAIN_REVISION=${config.CurrentBuildCommit}",
```
Could be Optional -> use Current COMMIT ID if not overridden

wbeard replied:
Is there an env var in Jenkins I can default to here if this arg's not passed in? I could do it further down in the script, but figured if a smart default's available in this context, we should do it here to keep the script as simple as possible.

sbhaskara-ca commented on resources/postDeployAndCommitsToInsights.sh:
```diff
@@ -0,0 +1,255 @@
+#!/bin/bash
+
+echo "API_URL: $API_URL"
```
Example for API URL? What is the format?

sbhaskara-ca commented on resources/updateDeployToInsights.sh:
```diff
@@ -0,0 +1,146 @@
+
+    if [[ "$OSTYPE" == "darwin"* ]]; then
+        # macOS
+        date -r $seconds -u +"%Y-%m-%dT%H:%M:%S%z"
```
Ensure the timezone is correct, as Rally expects the dates in UTC; we may need to use `Z` instead of `z`

wbeard replied:
I'll need to dig into this more!

sbhaskara-ca commented on resources/updateDeployToInsights.sh:
```diff
@@ -0,0 +1,146 @@
+query_deploy_id() {
+    local build_id=$1
+    local response
+    response=$(curl -s -H "ZSESSIONID: $API_KEY" "$API_URL/vsmdeploy?query=(BuildId%20=%20$build_id)&workspace=workspace/$API_WORKSPACE_OID&fetch=ObjectID")
```
expect `API_URL` will only be either `https://rally1.rallydev.com` or `https:\\vsi.rallydev.com`

sbhaskara-ca commented on resources/postDeployAndCommitsToInsights.sh:
```diff
@@ -0,0 +1,255 @@
+while IFS= read -r line; do
+    # Read the line
+    read -r commit_id timestamp <<< "$line"
+
```
Add some validation on the commit_id as sometimes the repo's that are polled might not have any commits done on a repo.

sbhaskara-ca reviewed (changes requested):
Add Readme with documentation and a sample Jenkins pipeline as example on the usage of the steps

wbeard committed:
Add example Jenkinsfile

wbeard committed:
Add docs about the methods being shared in jenkins

wbeard committed:
Shorten config variable name

wbeard committed:
Address most PR Feedback

- Make deploy finished optional for postDeployAndCommitsToInsights
- Make deploy is successful optional for postDeployAndCommitsToInsights
- Dynamically build json payload in case the optional fields are not available
- Shorten required string for API URL
- Fix date parsing to not include time zone since it will be written down at the correct instant and always read out as the workspace timezone
- Handle case where commit_log may be empty

wbeard committed:
Better example url in Jenkinsfile

wbeard committed:
Fix WorkspaceObjectId config param in Jenkinsfile

wbeard committed:
try a different empty check

wbeard committed:
Clean up

- Shorter names
- More defaults making less required args
- Fix date formatting

wbeard committed:
Add finding a component for a deploy by component name

wbeard committed:
Add more thorough documentation. Make more defaults for configs.

sbhaskara-ca commented:
@wbeard this looks good. If you have no more planned changes; I will proceed to merge and tag this.

wbeard commented:
> @wbeard this looks good. If you have no more planned changes; I will proceed to merge and tag this.

I made a few changes last night that I'm going to test out in my environment. I'll comment back once I've tested them and the branch is in a stable place!

wbeard committed:
Change up strategy for defaulting values

wbeard committed:
Update README

wbeard commented:
@sbhaskara-ca I went ahead and implemented the "lookup component by component name" path we talked about. All appears functional in my testing. I am done with the PR on my end, so feel free to merge!
